File: api/remote_source.go
package api

import (
	"context"
	"fmt"
	"net/url"
	"os/exec"
	"path"
	"sort"
	"strings"
	"time"
)

// ParsedRemoteSource is the normalized result of parsing any remote URL format.
// It represents a runbook located in a remote git repository.
type ParsedRemoteSource struct {
	Host     string // "github.com" or "gitlab.com"
	Owner    string // "gruntwork-io"
	Repo     string // "runbooks"
	Ref      string // "main", "v1.0" — empty if not yet resolved
	Path     string // "runbooks/setup-vpc" — path within the repo to the runbook dir
	CloneURL string // "https://github.com/gruntwork-io/runbooks.git"

	// IsBlobURL is true when the URL used /blob/ instead of /tree/.
	// This means the path points to a file (e.g., runbook.mdx) rather than a directory,
	// and ResolveRef should adjust the path to the parent directory.
	IsBlobURL bool

	// rawRefAndPath is the unresolved ref+path segment from browser URLs.
	// For browser URLs like /tree/main/path, this is "main/path" and needs
	// ResolveRef() to split it into Ref and Path.
	// Empty for OpenTofu-style URLs where ref and path are already separated.
	rawRefAndPath string
}

// NeedsRefResolution returns true if the ref has not been resolved yet
// (browser URLs where ref is embedded in the path).
func (p *ParsedRemoteSource) NeedsRefResolution() bool {
	return p.rawRefAndPath != "" && p.Ref == ""
}

// RawRefAndPath returns the unresolved ref+path segment for browser URLs.
func (p *ParsedRemoteSource) RawRefAndPath() string {
	return p.rawRefAndPath
}

// Resolve separates the git ref from the repo path when they're ambiguous.
//
// Browser URLs like github.com/org/repo/tree/main/infra/vpc embed the ref and
// path together as "main/infra/vpc". This could mean ref="main" path="infra/vpc",
// or ref="main/infra" path="vpc" if a branch named "main/infra" exists. Resolve
// uses git ls-remote to find the correct split.
//
// For blob URLs (pointing to a file rather than a directory), the path is also
// adjusted to the parent directory, since runbooks live in directories.
//
// token is used to authenticate the ls-remote call if non-empty.
func (p *ParsedRemoteSource) Resolve(token string) error {
	if p.NeedsRefResolution() {
		cloneURL := InjectGitToken(p.CloneURL, token)
		ref, repoPath := ResolveRef(cloneURL, p.RawRefAndPath(), p.IsBlobURL)
		p.Ref = ref
		p.Path = repoPath
	} else if p.IsBlobURL && p.Path != "" {
		p.Path = AdjustBlobPath(p.Path)
	}
	return nil
}

// ParseRemoteSource parses any supported remote URL format into a normalized form.
// Returns nil, nil if the input is not a remote source (i.e., it's a local path).
// Returns nil, error if it looks like a remote source but is malformed.
//
// Supported formats:
//   - git::https://host/owner/repo.git//path?ref=v1.0  (OpenTofu git:: prefix)
//   - github.com/owner/repo//path?ref=v1.0             (OpenTofu GitHub shorthand)
//   - https://github.com/owner/repo/tree/ref/path      (GitHub browser URL)
//   - https://github.com/owner/repo/blob/ref/file.mdx  (GitHub blob URL)
//   - https://gitlab.com/owner/repo/-/tree/ref/path    (GitLab browser URL)
//   - https://gitlab.com/owner/repo/-/blob/ref/file.mdx (GitLab blob URL)
//   - https://github.com/owner/repo                     (plain repo URL)
func ParseRemoteSource(raw string) (*ParsedRemoteSource, error) {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return nil, nil
	}

	lowered := strings.ToLower(raw)

	// 1. git::https:// or git::http:// prefix (OpenTofu generic git source)
	if strings.HasPrefix(lowered, "git::https://") || strings.HasPrefix(lowered, "git::http://") {
		return parseGitPrefixURL(raw)
	}

	// 2. github.com/ without scheme (OpenTofu GitHub shorthand)
	if strings.HasPrefix(lowered, "github.com/") {
		return parseGitHubShorthand(raw)
	}

	// 3. https://github.com/ or http://github.com/ (GitHub browser URL)
	if strings.HasPrefix(lowered, "https://github.com/") || strings.HasPrefix(lowered, "http://github.com/") {
		return parseBrowserURL(raw, "github.com")
	}

	// 4. https://gitlab.com/ or http://gitlab.com/ (GitLab browser URL)
	if strings.HasPrefix(lowered, "https://gitlab.com/") || strings.HasPrefix(lowered, "http://gitlab.com/") {
		return parseBrowserURL(raw, "gitlab.com")
	}

	// Not a recognized remote source — treat as local path
	return nil, nil
}

// --- OpenTofu-style URL parsers (git:: prefix and GitHub shorthand) ---

// parseGitPrefixURL parses an OpenTofu git::https:// URL.
// Format: git::https://host/owner/repo.git//path?ref=v1.0
func parseGitPrefixURL(raw string) (*ParsedRemoteSource, error) {
	rawURL := strings.TrimPrefix(raw, "git::")

	parsed, err := url.Parse(rawURL)
	if err != nil {
		return nil, fmt.Errorf("invalid git URL %q: %w", raw, err)
	}

	host := strings.ToLower(parsed.Hostname())
	if host == "" {
		return nil, fmt.Errorf("invalid git URL %q: missing host", raw)
	}

	owner, repo, subPath, err := splitTofuSourcePath(parsed.Path, raw, "expected owner/repo in path")
	if err != nil {
		return nil, err
	}

	return &ParsedRemoteSource{
		Host:     host,
		Owner:    owner,
		Repo:     repo,
		Ref:      parsed.Query().Get("ref"),
		Path:     subPath,
		CloneURL: fmt.Sprintf("https://%s/%s/%s.git", host, owner, repo),
	}, nil
}

// parseGitHubShorthand parses an OpenTofu GitHub shorthand URL.
// Format: github.com/owner/repo//path?ref=v1.0
func parseGitHubShorthand(raw string) (*ParsedRemoteSource, error) {
	parsed, err := url.Parse("https://" + raw)
	if err != nil {
		return nil, fmt.Errorf("invalid GitHub source %q: %w", raw, err)
	}

	owner, repo, subPath, err := splitTofuSourcePath(parsed.Path, raw, "expected github.com/owner/repo")
	if err != nil {
		return nil, err
	}

	return &ParsedRemoteSource{
		Host:     "github.com",
		Owner:    owner,
		Repo:     repo,
		Ref:      parsed.Query().Get("ref"),
		Path:     subPath,
		CloneURL: fmt.Sprintf("https://github.com/%s/%s.git", owner, repo),
	}, nil
}

// splitTofuSourcePath extracts owner, repo, and sub-path from a URL path
// that uses the OpenTofu // separator convention (e.g., /owner/repo.git//sub/path).
func splitTofuSourcePath(urlPath, rawInput, errHint string) (owner, repo, subPath string, err error) {
	fullPath := strings.TrimPrefix(urlPath, "/")
	fullPath = strings.TrimSuffix(fullPath, "/")

	var repoPath string
	if idx := strings.Index(fullPath, "//"); idx >= 0 {
		repoPath = fullPath[:idx]
		subPath = strings.TrimPrefix(fullPath[idx+2:], "/")
		subPath = strings.TrimSuffix(subPath, "/")
	} else {
		repoPath = fullPath
	}

	repoPath = strings.TrimSuffix(repoPath, ".git")
	parts := strings.SplitN(repoPath, "/", 3)
	if len(parts) < 2 || parts[0] == "" || parts[1] == "" {
		return "", "", "", fmt.Errorf("invalid source %q: %s", rawInput, errHint)
	}

	return parts[0], parts[1], subPath, nil
}

// --- Browser URL parser (GitHub and GitLab) ---

// parseBrowserURL parses a GitHub or GitLab browser URL.
//
// GitHub formats:
//   - https://github.com/owner/repo/tree/ref/path
//   - https://github.com/owner/repo/blob/ref/path
//   - https://github.com/owner/repo
//
// GitLab formats:
//   - https://gitlab.com/owner/repo/-/tree/ref/path
//   - https://gitlab.com/owner/repo/-/blob/ref/path
//   - https://gitlab.com/owner/repo
func parseBrowserURL(raw string, host string) (*ParsedRemoteSource, error) {
	parsed, err := url.Parse(raw)
	if err != nil {
		return nil, fmt.Errorf("invalid %s URL %q: %w", host, raw, err)
	}

	pathStr := strings.TrimPrefix(parsed.Path, "/")
	pathStr = strings.TrimSuffix(pathStr, "/")

	segments := strings.Split(pathStr, "/")
	if len(segments) < 2 {
		return nil, fmt.Errorf("invalid %s URL %q: expected %s/owner/repo", host, raw, host)
	}

	owner := segments[0]
	repo := strings.TrimSuffix(segments[1], ".git")
	if owner == "" || repo == "" {
		return nil, fmt.Errorf("invalid %s URL %q: empty owner or repo", host, raw)
	}

	result := &ParsedRemoteSource{
		Host:     host,
		Owner:    owner,
		Repo:     repo,
		CloneURL: fmt.Sprintf("https://%s/%s/%s.git", host, owner, repo),
	}

	// Extract the tree/blob action and rawRefAndPath from remaining segments.
	// GitLab uses /-/tree/ while GitHub uses /tree/ directly.
	action, refAndPath, err := extractBrowserAction(segments[2:], host, raw)
	if err != nil {
		return nil, err
	}
	if action == "blob" {
		result.IsBlobURL = true
	}
	result.rawRefAndPath = refAndPath

	return result, nil
}

// extractBrowserAction parses the segments after owner/repo to find the tree/blob
// action and the ref+path string. Returns empty strings if there's no action.
func extractBrowserAction(segments []string, host, raw string) (action, refAndPath string, err error) {
	if len(segments) == 0 {
		return "", "", nil
	}

	// GitLab uses /-/ before tree/blob; skip it
	if segments[0] == "-" {
		segments = segments[1:]
		if len(segments) == 0 {
			return "", "", nil
		}
	}

	action = segments[0]
	switch action {
	case "tree", "blob":
		remaining := segments[1:]
		if len(remaining) == 0 {
			return "", "", fmt.Errorf("invalid %s URL %q: missing ref after /%s/", host, raw, action)
		}
		return action, strings.Join(remaining, "/"), nil
	default:
		// For GitHub: unexpected third segment like /settings or /actions
		// For GitLab: unknown action after /-/
		if host == "github.com" {
			return "", "", fmt.Errorf("invalid %s URL %q: unexpected path segment /%s/", host, raw, action)
		}
		// GitLab: treat unknown /-/<action> as plain repo URL
		return "", "", nil
	}
}

// --- Ref resolution ---

// ResolveRef uses git ls-remote to resolve the ref from a browser URL's ambiguous
// ref+path segment (e.g., "main/runbooks/setup-vpc" could be ref="main" path="runbooks/setup-vpc"
// or ref="main/runbooks" path="setup-vpc" if a branch named "main/runbooks" exists).
//
// cloneURL should already have the token injected if auth is needed.
// rawRefAndPath is the segment after /tree/ or /blob/ in the browser URL.
// isBlobURL indicates whether the original URL used /blob/ (path points to a file).
//
// Returns the resolved ref name and the remaining path within the repo.
// Always succeeds: if ls-remote fails or no ref matches, falls back to the first path segment.
func ResolveRef(cloneURL, rawRefAndPath string, isBlobURL bool) (ref string, repoPath string) {
	if rawRefAndPath == "" {
		return "", ""
	}

	// Get all refs from the remote
	refs, err := listRemoteRefsFn(cloneURL)
	if err != nil {
		return resolveRefFallback(rawRefAndPath, isBlobURL)
	}

	// Sort refs by length descending so we find the longest match first
	sort.Slice(refs, func(i, j int) bool {
		return len(refs[i]) > len(refs[j])
	})

	// Find the longest ref that matches the beginning of rawRefAndPath
	for _, r := range refs {
		if rawRefAndPath == r {
			return r, ""
		}
		if strings.HasPrefix(rawRefAndPath, r+"/") {
			remaining := strings.TrimPrefix(rawRefAndPath, r+"/")
			remaining = strings.TrimSuffix(remaining, "/")
			if isBlobURL {
				remaining = AdjustBlobPath(remaining)
			}
			return r, remaining
		}
	}

	return resolveRefFallback(rawRefAndPath, isBlobURL)
}

// resolveRefFallback splits rawRefAndPath using the first segment as the ref.
// This never fails — it always produces a best-effort split.
func resolveRefFallback(rawRefAndPath string, isBlobURL bool) (string, string) {
	parts := strings.SplitN(rawRefAndPath, "/", 2)
	ref := parts[0]
	repoPath := ""
	if len(parts) > 1 {
		repoPath = strings.TrimSuffix(parts[1], "/")
	}
	if isBlobURL {
		repoPath = AdjustBlobPath(repoPath)
	}
	return ref, repoPath
}

// AdjustBlobPath adjusts a path from a /blob/ URL to point to the parent directory.
// For example, "runbooks/setup-vpc/runbook.mdx" becomes "runbooks/setup-vpc".
// Uses path.Dir (POSIX) intentionally since these are URL/repo paths, not filesystem paths.
func AdjustBlobPath(p string) string {
	if p == "" {
		return ""
	}
	dir := path.Dir(p)
	if dir == "." {
		return ""
	}
	return dir
}

// listRemoteRefsFn is the function used to list remote refs.
// It is a variable so tests can inject a fake implementation.
var listRemoteRefsFn = listRemoteRefsImpl

// listRemoteRefsImpl runs git ls-remote and returns a list of ref names
// (with refs/heads/ and refs/tags/ prefixes stripped).
func listRemoteRefsImpl(cloneURL string) ([]string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, "git", "ls-remote", "--heads", "--tags", "--refs", cloneURL)
	output, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("git ls-remote failed: %w", err)
	}

	var refs []string
	for _, line := range strings.Split(string(output), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		// Format: <sha>\t<ref>
		parts := strings.SplitN(line, "\t", 2)
		if len(parts) != 2 {
			continue
		}
		refName := parts[1]
		refName = strings.TrimPrefix(refName, "refs/heads/")
		refName = strings.TrimPrefix(refName, "refs/tags/")
		refs = append(refs, refName)
	}

	return refs, nil
}

File: src/remote-source.ts
/**
 * Remote URL parsing.
 *
 * Handles OpenTofu-style git:: URLs, GitHub/GitLab shorthand, and browser URLs.
 */
import { Effect, Stream } from "effect"
import { ProcessSpawner } from "./services/ProcessSpawner.ts"
import type { SpawnError } from "./errors/index.ts"
import { GitError, RemoteSourceError } from "./errors/index.ts"
import { gitSpawnEnv } from "./domain/git/env.ts"
import { redactSecrets } from "./domain/vcs/redact.ts"
import { isGitLabHost } from "./domain/git/gitlab-host.ts"
import type { ParsedRemoteSource } from "./types.ts"

// ---------------------------------------------------------------------------
// URL patterns
// ---------------------------------------------------------------------------

/**
 * OpenTofu git:: prefix: git::https://host/owner/.../repo.git//path?ref=v1.0
 * The owner/repo portion (everything between the host and the `//` path
 * delimiter) may be a nested group path on GitLab.
 */
const GIT_PREFIX_REGEX = /^git::/i

/**
 * OpenTofu shorthand: github.com/owner/repo//path?ref=v1.0 (or gitlab.com,
 * where the owner may be a nested group path). Scheme-less, so it is given
 * an https:// scheme before parsing.
 */
const SHORTHAND_REGEX = /^(?:github|gitlab)\.com\//i

/** GitHub browser tree URL: https://github.com/owner/repo/tree/ref/path */
const GITHUB_TREE_REGEX =
  /^https?:\/\/github\.com\/([^/]+)\/([^/]+)\/tree\/(.+)$/

/** GitHub browser blob URL: https://github.com/owner/repo/blob/ref/file */
const GITHUB_BLOB_REGEX =
  /^https?:\/\/github\.com\/([^/]+)\/([^/]+)\/blob\/(.+)$/

/**
 * GitLab browser tree URL: https://host/group/.../repo/-/tree/ref/path
 * The `/-/` marker is GitLab-specific, so the host may be gitlab.com or a
 * self-hosted instance, and the owner may be a nested group path.
 */
const GITLAB_TREE_REGEX =
  /^https?:\/\/([^/]+)\/(.+?)\/-\/tree\/(.+)$/

/** GitLab browser blob URL: https://host/group/.../repo/-/blob/ref/file */
const GITLAB_BLOB_REGEX =
  /^https?:\/\/([^/]+)\/(.+?)\/-\/blob\/(.+)$/

/**
 * Plain GitHub repo URL: https://github.com/owner/repo (no nested groups).
 * The repo name may contain dots (e.g. `docs.example.io`).
 */
const PLAIN_GITHUB_REPO_REGEX =
  /^https?:\/\/github\.com\/([^/]+)\/([^/]+?)(?:\.git)?$/

/**
 * Plain GitLab repo URL: https://<host>/group/.../repo
 * Captures the host so self-hosted instances are supported; the caller only
 * accepts it when the host is recognizably GitLab (isGitLabHost). Supports
 * nested groups — the last path segment is the repo (project) and everything
 * before it is the owner.
 */
const PLAIN_GITLAB_REPO_REGEX =
  /^https?:\/\/([^/]+)\/(.+?)(?:\.git)?$/

/**
 * Split a slash-delimited `owner/.../repo` path into its owner and repo parts.
 * The last segment is the repo (project, with any `.git` suffix stripped) and
 * everything before it is the owner. For GitLab nested groups the owner is the
 * full group path, e.g. `group/subgroup/project` → owner "group/subgroup",
 * repo "project".
 */
const splitOwnerRepo = (
  ownerRepoPath: string,
): { owner: string; repo: string } => {
  const segments = ownerRepoPath.split("/").filter(Boolean)
  const repo = (segments.pop() ?? "").replace(/\.git$/, "")
  const owner = segments.join("/")
  return { owner, repo }
}

/** Remove leading and trailing slashes. */
const trimSlashes = (s: string): string => s.replace(/^\/+|\/+$/g, "")

// ---------------------------------------------------------------------------
// parseRemoteSource
// ---------------------------------------------------------------------------

export const parseRemoteSource = (raw: string): Effect.Effect<ParsedRemoteSource, RemoteSourceError> =>
  Effect.gen(function* () {
    const trimmed = raw.trim()
    if (!trimmed) {
      return yield* Effect.fail(new RemoteSourceError({ url: raw, message: "empty URL" }))
    }
    const unsupported = new RemoteSourceError({ url: raw, message: "unsupported URL format" })

    // Normalize once, as the golang parser did with url.Parse: strip the
    // OpenTofu `git::` prefix, give the scheme-less shorthand a scheme, then
    // parse. Every form below reads only the lowercased host, the decoded
    // path and (OpenTofu forms) the `ref` query parameter, so a ?query or
    // #fragment never leaks into the repo, ref or path.
    const isGitPrefixed = GIT_PREFIX_REGEX.test(trimmed)
    const isShorthand = !isGitPrefixed && SHORTHAND_REGEX.test(trimmed)
    const input = isGitPrefixed
      ? trimmed.replace(GIT_PREFIX_REGEX, "")
      : isShorthand
        ? `https://${trimmed}`
        : trimmed
    const url = yield* Effect.try({ try: () => new URL(input), catch: () => unsupported })
    if (url.protocol !== "https:" && url.protocol !== "http:") {
      return yield* Effect.fail(unsupported)
    }
    // `host` keeps a non-default port (a self-hosted instance on :8443).
    const host = url.host
    // `new URL` percent-encodes the path (spaces, non-ASCII); decode it back
    // so it names the directory in the repo. decodeURI leaves `%2F` encoded:
    // `new URL` has already resolved `.`/`..` segments, and decoding a slash
    // would let `..%2F..` rebuild one that escapes the clone directory.
    const pathname = yield* Effect.try(() => decodeURI(url.pathname)).pipe(
      Effect.orElseSucceed(() => url.pathname),
    )
    // `new URL` only resolves `/`-delimited dot segments, but decodeURI turns
    // `%5C` into a backslash, which Windows' path.join treats as a separator:
    // `..%5C..%5Cetc` would climb out of the clone there. No repo, ref or
    // path legitimately has a `..` segment, so reject one under either
    // separator (on every platform, so the check is tested everywhere).
    if (pathname.split(/[\\/]/).includes("..")) {
      return yield* Effect.fail(unsupported)
    }

    // 1) OpenTofu forms: git::https://host/owner/.../repo.git//path?ref=v1.0
    //    and the github.com / gitlab.com shorthand. The `//path` subdir is
    //    optional (the runbook is then at the repo root), and only `ref` is
    //    read from the query — OpenTofu sources may carry others (`depth`).
    if (isGitPrefixed || isShorthand) {
      const fullPath = trimSlashes(pathname)
      const sep = fullPath.indexOf("//")
      const repoPath = sep >= 0 ? fullPath.slice(0, sep) : fullPath
      const path = sep >= 0 ? trimSlashes(fullPath.slice(sep + 2)) || undefined : undefined
      const { owner, repo } = splitOwnerRepo(repoPath)
      // GitHub has no nested groups → exactly owner/repo. GitLab reserves the
      // `-` segment for its own routes, so a repo path containing one is a
      // browser URL (`gitlab.com/g/p/-/tree/main/x`), not a nested group.
      if (
        !owner ||
        !repo ||
        (host === "github.com" && owner.includes("/")) ||
        repoPath.split("/").includes("-")
      ) {
        return yield* Effect.fail(unsupported)
      }
      return {
        host,
        owner,
        repo,
        ref: url.searchParams.get("ref") || undefined,
        path,
        cloneURL: `https://${host}/${owner}/${repo}.git`,
        isBlobURL: false,
      }
    }

    // Browser and plain repo URLs are matched on host + path alone, without
    // a trailing slash.
    const normalized = `https://${host}${pathname.replace(/\/+$/, "")}`

    // 2) GitHub tree URL
    let match = normalized.match(GITHUB_TREE_REGEX)
    if (match) {
      const [, owner, repo, refAndPath] = match
      return {
        host: "github.com",
        owner,
        repo,
        // ref/path split is ambiguous; set path as combined and resolve later
        path: refAndPath,
        refInPath: true,
        cloneURL: `https://github.com/${owner}/${repo}.git`,
        isBlobURL: false,
      }
    }

    // 3) GitHub blob URL
    match = normalized.match(GITHUB_BLOB_REGEX)
    if (match) {
      const [, owner, repo, refAndPath] = match
      return {
        host: "github.com",
        owner,
        repo,
        path: refAndPath,
        refInPath: true,
        cloneURL: `https://github.com/${owner}/${repo}.git`,
        isBlobURL: true,
      }
    }

    // 4) GitLab tree URL
    match = normalized.match(GITLAB_TREE_REGEX)
    if (match) {
      const [, host, ownerRepoPath, refAndPath] = match
      const { owner, repo } = splitOwnerRepo(ownerRepoPath)
      return {
        host,
        owner,
        repo,
        path: refAndPath,
        refInPath: true,
        cloneURL: `https://${host}/${owner}/${repo}.git`,
        isBlobURL: false,
      }
    }

    // 5) GitLab blob URL
    match = normalized.match(GITLAB_BLOB_REGEX)
    if (match) {
      const [, host, ownerRepoPath, refAndPath] = match
      const { owner, repo } = splitOwnerRepo(ownerRepoPath)
      return {
        host,
        owner,
        repo,
        path: refAndPath,
        refInPath: true,
        cloneURL: `https://${host}/${owner}/${repo}.git`,
        isBlobURL: true,
      }
    }

    // 6) Plain GitHub repo URL (GitHub has no nested groups → exactly owner/repo)
    match = normalized.match(PLAIN_GITHUB_REPO_REGEX)
    if (match) {
      const [, owner, repo] = match
      return {
        host: "github.com",
        owner,
        repo,
        cloneURL: `https://github.com/${owner}/${repo}.git`,
        isBlobURL: false,
      }
    }

    // 7) Plain GitLab repo URL (supports nested groups → last segment is the
    //    repo). Accepts gitlab.com and self-hosted GitLab hosts recognizable by
    //    name; other hosts (e.g. bitbucket.org) fall through to "unsupported".
    match = normalized.match(PLAIN_GITLAB_REPO_REGEX)
    if (match) {
      const [, host, ownerRepoPath] = match
      if (isGitLabHost(host)) {
        const { owner, repo } = splitOwnerRepo(ownerRepoPath)
        // A GitLab project always lives under at least one namespace, so a
        // single-segment path (no owner) is not a valid repo URL.
        if (owner) {
          return {
            host,
            owner,
            repo,
            cloneURL: `https://${host}/${owner}/${repo}.git`,
            isBlobURL: false,
          }
        }
      }
    }

    return yield* Effect.fail(unsupported)
  })

// ---------------------------------------------------------------------------
// needsRefResolution
// ---------------------------------------------------------------------------

/**
 * Returns true for browser-style URLs where the ref/path boundary is ambiguous
 * (e.g. `/tree/main/some/path` — is the ref "main" or "main/some"?).
 */
export function needsRefResolution(parsed: ParsedRemoteSource): boolean {
  // Browser URLs store the combined ref+path in `path` without a separate `ref`.
  // An OpenTofu-style URL's `//path` is only ever a path, with or without an
  // explicit `?ref` (no ref means the default branch).
  return parsed.refInPath === true && parsed.ref === undefined && parsed.path !== undefined
}

// ---------------------------------------------------------------------------
// resolveRef
// ---------------------------------------------------------------------------

/**
 * Uses `git ls-remote` to determine the correct ref from a combined ref/path string.
 * Tries longest match first so that a ref like "feature/foo" beats "feature".
 *
 * A failed ls-remote (auth, network, missing repo) fails with a GitError
 * rather than guessing a ref, so the caller can classify it like a failed
 * clone. A segment that matches no ref falls back to being the ref itself,
 * which is how a commit SHA (a permalink) resolves.
 */
export const resolveRef = (
  cloneURL: string,
  rawRefAndPath: string,
): Effect.Effect<
  { ref: string; path: string | undefined },
  GitError | SpawnError,
  ProcessSpawner
> =>
  Effect.gen(function* () {
    const spawner = yield* ProcessSpawner

    // Fetch all remote refs. gitSpawnEnv keeps ssh non-interactive so an
    // ls-remote against an unknown SSH host fails fast instead of hanging on
    // the host-key prompt.
    const proc = yield* spawner.spawn("git", ["ls-remote", "--refs", cloneURL], {
      env: gitSpawnEnv(),
    })
    const lines: string[] = []
    const stderrLines: string[] = []
    yield* Stream.runForEach(proc.output, (line) => {
      if (line.source === "stderr") {
        stderrLines.push(line.line)
      } else if (line.line.trim()) {
        lines.push(line.line)
      }
      return Effect.void
    })
    const code = yield* proc.exitCode
    if (code !== 0) {
      // The URL carries the token, so scrub it from git's output.
      return yield* Effect.fail(
        new GitError({
          command: "git ls-remote",
          stderr: redactSecrets(stderrLines.join("\n")),
          exitCode: code,
        }),
      )
    }

    // Build set of known ref names (strip refs/heads/ and refs/tags/)
    const knownRefs = new Set<string>()
    for (const line of lines) {
      const parts = line.split("\t")
      if (parts.length >= 2) {
        const refName = parts[1]
          .trim()
          .replace(/^refs\/heads\//, "")
          .replace(/^refs\/tags\//, "")
        knownRefs.add(refName)
      }
    }

    // Split rawRefAndPath into segments and try longest ref match first
    const segments = rawRefAndPath.split("/")
    for (let i = segments.length; i >= 1; i--) {
      const candidateRef = segments.slice(0, i).join("/")
      if (knownRefs.has(candidateRef)) {
        const remainingPath = segments.slice(i).join("/") || undefined
        return { ref: candidateRef, path: remainingPath }
      }
    }

    // Fall back: assume first segment is the ref
    const ref = segments[0]
    const path = segments.slice(1).join("/") || undefined
    return { ref, path }
  })

// ---------------------------------------------------------------------------
// adjustBlobPath
// ---------------------------------------------------------------------------

/**
 * Converts a blob path to its parent directory so the tool fetches the
 * containing folder rather than a single file.
 */
export function adjustBlobPath(parsed: ParsedRemoteSource): ParsedRemoteSource {
  if (!parsed.isBlobURL || !parsed.path) return parsed
  const lastSlash = parsed.path.lastIndexOf("/")
  const adjustedPath = lastSlash > 0 ? parsed.path.substring(0, lastSlash) : undefined
  return { ...parsed, path: adjustedPath, isBlobURL: false }
}
